[FLINK-16512][task] Unaligned checkpoints: API for persistence

rkhachatryan opened the pull request:
## What is the purpose of the change

*(Add an API to save and restore channel state for the unaligned checkpoints)*


## Brief change log

  - *added info classes for InputChannel and ResultSubpartition*
  - *added info corresponding state handles (interfeces)*
  - *added info Reader and Writer interfaces*
  - *added NO_OP implementation*


## Verifying this change

This change is only an API part without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateReader.java:
```diff
@@ -0,0 +1,59 @@
+	/**
+	 * ReadResult.
+	 */
+	enum ReadResult {HAS_MORE_DATA, NO_MORE_DATA}
```
nit: add spaces before and after values

AHeise replied:
Do we need an enum (vs. a boolean)? Are you planning to expand the values?

rkhachatryan replied:
Currently, I'm not planning to extend it. 
It's just more readable than a boolean.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateReader.java:
```diff
@@ -0,0 +1,59 @@
+	ReadResult readOutputData(ResultSubpartitionInfo info, Buffer buffer);
+
+	@Override
+	default void close() throws Exception {
```
How many implementation with no-op close do you expect? I'd assume it's only one and then I'd avoid having it as default implementation.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriter.java:
```diff
@@ -0,0 +1,101 @@
+	 * <p>Must be called after {@link #start} and before {@link #finish(long)}.</p>
+	 * @param startSeqNum is the sequence number of the 1st passed buffer
+	 *                    It’s intended use is incremental snapshots
+	 *                    For the restored buffers it should be -1
```
I'd don't get the idea of having -1 here. Doesn't it make usage more complicated?
I thought adding the real sequence number would make incremental checkpointing easier (you just need to remember the highest number of the previous checkpoint).

rkhachatryan replied:
For incremental checkpointing, we need to distinguish somehow restored buffers from new buffers. Given that:
* sequence numbers currently start from 0 after the restart
* buffers don't carry any sequence numbers themselves
we can't compare the given argument with something from the previous run.

Can you please explain what are "real sequence numbers"?

AHeise replied:
I thought about storing the sequence numbers alongside the data. But I haven't thought about sequence numbers being reset.
I'm simply not sure how incremental checkpoints should work if we don't have that information. 

Let's assume there is a checkpoint where for one specific channel buffer 0..9 are stored. Now, there is a recovery where only 0..5 are completely unspilled before a new checkpoint is taken, in which 6..15 are stored. For incremental snapshot, we would simply skip storing 6..9 because at the current API, they'd get a -1. However, how would the recovery know where to resume?

rkhachatryan replied:
I assume that by unspilling you mean sending the buffers out (to the operator or to downstream); and that we have 6 new buffers.

The other one is to count how many buffers or bytes actually are in-flight and store this offset.
One option would be to do a full snapshot if there any negative sequence numbers.

Of course, having SNs inside the buffers would simplify a lot, but I guess it would require more effort and discussion.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriter.java:
```diff
@@ -0,0 +1,101 @@
+	 * @param startSeqNum is the sequence number of the 1st passed buffer
+	 *                    It’s intended use is incremental snapshots
+	 *                    For the restored buffers it should be -1
+	 *                    For MVP it can always be -2
```
I'd probably not refer to MVP in the code. Noone from outside will know what it means. 
How about `Can be -2 if unknown, which will fail on incremental checkpoint`.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriter.java:
```diff
@@ -0,0 +1,101 @@
+
+	/**
+	 * Add in-flight buffers from the {@link org.apache.flink.runtime.io.network.partition.consumer.InputChannel InputChannel}.
+	 * <p>Must be called after {@link #start} and before {@link #finish(long)}.</p>
```
start(long)

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriter.java:
```diff
@@ -0,0 +1,101 @@
+	void addInputData(long checkpointId, InputChannelInfo info, int startSeqNum, Buffer... data);
+
+	/**
+	 * Add in-flight buffers from the {@link org.apache.flink.runtime.io.network.partition.ResultSubpartition ResultSubpartition}.
```
See comments about `addInputData`.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriter.java:
```diff
@@ -0,0 +1,101 @@
+	/**
+	 * <p>Must be called after {@link #finish}.</p>
+	 */
+	Future<Collection<StateObject>> getWriteCompletionFuture(long checkpointId);
```
Could we merge that with finish?
Do you really want to use Future instead of CompletableFuture?

rkhachatryan replied:
I'm not sure if these calls will come from the same place.
I think this can be refined once we get closer to state injection implementation.

zhijiangW replied:
I guess it probably happens in different places. Anyway, it might be no harm to separate the different semantics.

rkhachatryan replied:
Agree.
I will also change the contract slightly so this method should be called after `start()`.

AHeise replied:
Ah yes, if it does not need to be called after finish, it will be much more usable and I agree it should be separate.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriter.java:
```diff
@@ -0,0 +1,101 @@
+	Future<Collection<StateObject>> getWriteCompletionFuture(long checkpointId);
+
+	@Override
+	default void close() throws Exception {
```
as before: How many implementation with no-op close do you expect? I'd assume it's only one and then I'd avoid having it as default implementation.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/InputChannelInfo.java:
```diff
@@ -0,0 +1,46 @@
+ * Identifies {@link org.apache.flink.runtime.io.network.partition.consumer.InputChannel} in a given subtask.
+ */
+@Internal
+public class InputChannelInfo implements Serializable {
```
It's confusing that we now have `InputChannelID` and `InputChannelInfo`. I don't see a good solution though. Similarly, it's a bit of a pity that we cannot reuse this class for `ResultSubpartitionInfo` (virtually it's the same thing). Maybe @zhijiangW has an idea.

rkhachatryan replied:
Unfortunately, we can't use `ID`s as they are currently assigned at random. I'll put it into the javadoc.

To identify state, we logically need: `source task id`, `target task id`, `isUpstream`. Instead of `isUpstream`, I thought it's better to have an explicit entity for both `InputChannel` and `ResultSubpartition`.

Regarding names, it was actually @zhijiangW (good) idea:)

zhijiangW replied:
It indeed brings confusing for having both `InputChannelID` and `InputChannelInfo` now.

`InputChannelID` is the unique id for identifying while requesting partitions from different tasks on upstream side.  
`InputChannelInfo` is a wrapper structure which would be easily extended to cover more infos (e.g. task index) future if necessary. And currently the strong requirement is to identify it for different `InputGate` instances.

The "inspiration" is form existing `TaskInfo` structure coincides with `executionId`. I also do not see a better solution yet.

rkhachatryan replied:
Let's stick with this structure then.

AHeise replied:
Ideally, we could replace `InputChannelID` with Roman's `InputChannelInfo`, but that's probably too much of an effort. It would certainly make programs easier to debug.

zhijiangW replied:
`InputChannelID` can not be replaced directly by InputChannelInfo` as I described the function above.

For every upstream task, it can receive the partition requests from the `InputChannel` of different task instances. And now it uses the global unique `InputChannelID` for distinguishing with different tasks.

But `InputChannelInfo` can only guarantee the uniqueness within one task scope,  and still has the same id among different tasks, unless we introduce `InputChannelID` into the `InputChannelInfo`.

rkhachatryan commented:
Thanks for the feedback. I've addressed some issues and rebased to current master.

CI failures are not related:
https://issues.apache.org/jira/browse/FLINK-16410
https://issues.apache.org/jira/browse/FLINK-16313

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriter.java:
```diff
@@ -0,0 +1,104 @@
+	 * Add in-flight buffers from the {@link org.apache.flink.runtime.io.network.partition.consumer.InputChannel InputChannel}.
+	 * <p>Must be called after {@link #start(long)} and before {@link #finish(long)}.</p>
+	 * @param startSeqNum is the sequence number of the 1st passed buffer.
+	 *                    It’s intended use is incremental snapshots.
```
nit: intended to use in

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriter.java:
```diff
@@ -0,0 +1,104 @@
+
+	/**
+	 * Add in-flight buffers from the {@link org.apache.flink.runtime.io.network.partition.consumer.InputChannel InputChannel}.
+	 * <p>Must be called after {@link #start(long)} and before {@link #finish(long)}.</p>
```
nit: remove </p>?

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateReader.java:
```diff
@@ -0,0 +1,62 @@
+	ReadResult readInputData(InputChannelInfo info, Buffer buffer);
+
+	/**
+	 * Put data into the supplied buffer to be injected into {@link org.apache.flink.runtime.io.network.partition.ResultSubpartition ResultSubpartition}.
```
ditto

rkhachatryan replied:
Sorry, didn't get you:)

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateReader.java:
```diff
@@ -0,0 +1,62 @@
+	enum ReadResult { HAS_MORE_DATA, NO_MORE_DATA }
+
+	/**
+	 * Put data into the supplied buffer to be injected into {@link org.apache.flink.runtime.io.network.partition.consumer.InputChannel InputChannel}.
```
nit: seems too long line. Can import the package before class or split the lines.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ResultSubpartitionInfo.java:
```diff
@@ -0,0 +1,48 @@
+
+/**
+ * Identifies {@link org.apache.flink.runtime.io.network.partition.ResultSubpartition ResultSubpartition} in a given subtask.
+ * Note that {@link org.apache.flink.runtime.jobgraph.IntermediateResultPartitionID IntermediateResultPartitionID}
```
Not only for random issue. `IntermediateResultPartitionID` is for identifying the whole `ResultPartition`, but we also need to identify the specific `ResultSubpartition` inside `ResultPartition.

rkhachatryan replied:
Yes, you're right.
I'll update the javadoc.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriter.java:
```diff
@@ -0,0 +1,104 @@
+	 * <p>Must be called after {@link #start(long)} and before {@link #finish(long)}.</p>
+	 * @param startSeqNum is the sequence number of the 1st passed buffer.
+	 *                    It’s intended use is incremental snapshots.
+	 *                    For the restored buffers it should be -1.
```
Btw, could we extract -1 and -2 as constants?

rkhachatryan replied:
Good idea!

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriter.java:
```diff
@@ -0,0 +1,116 @@
+public interface ChannelStateWriter extends AutoCloseable {
+
+	/**
+	 * Sequence number for restored buffers (rather than received from the upstream or from the operator).
```
I am a bit confused of the description `rather than received from the upstream or from the operator`. I may guess your meaning, but I wonder it might confuse other guys. E.g. for previously spilled input buffers the sequence number should also be -1 during recovery.

rkhachatryan replied:
I meant this constant for the not implemented case (MVP).
I'll clarify it in the javadocs.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriter.java:
```diff
@@ -0,0 +1,116 @@
+	int SEQUENCE_NUMBER_RESTORED = -1;
+
+	/**
+	 * Signifies that buffer sequence number is unknown.
```
TBH I am not very clear of the scenarios of unknown, can you explain a bit if possible in the javadoc?

rkhachatryan replied:
I meant this constant for the not implemented case (MVP).
I'll clarify it in the javadocs.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateReader.java:
```diff
@@ -0,0 +1,64 @@
+public interface ChannelStateReader extends AutoCloseable {
+
+	/**
+	 * ReadResult.
```
nit: ReadResult seems informal. Status of reading result

rkhachatryan commented:
I addressed remaining issues with the javadoc.
Thanks for reviewing @zhijiangW , @AHeise .
